# GTEx_HML2
Prepared by Aidan Burn
This contains all scripts used to analyze GTEx data for HML-2 expression
2 sets of Scripts are included here:
  1.Command line bash scripts for download, QC, alignment and counting
  2.R scripts for downstream normalization, data manipulation, differential expression and figure generation

The Bash scripts are run through the two overworld scripts,
  Overworld.sh runs the download and bamtofastq of 4 samples
    Overworld runs off of the Start_Variables.txt file
    It calls over_anvil_download to run the download of files from google cloud
    It calls over_wrapper_bamtofastq to run the fastq extraction from downloaded bam files
  Second_Overworld.sh runs the trim, QC, alignment and counting of each sample along with collecting all counts files for download
    Second_Overworld runs off the Variables.txt file
    It calls over_read_filter_quality.sbatch.sh to run Trimmomatic to trim reads and then run FastQC on trimmed files
    It calls over_alignment.sbatch.sh to run HiSat2 to align each fastq to HG38
    It calls over_Telescope.sbatch.sh to generate gene counts using Telescope
    It calls over_report.sh to run multiQC on all fastQC reports and collect all counts files in one directory
 Raw counts files are then generated
 
 The R scripts are used for data manipulation after count generation
 Telescope_Merge_Counts.R was used to collect sample count files organized by donor into tissue specific directories. Due to the usage of SUBJID by GTExV8 and SRAID by GTExV7 separate scripts were used
 
 HML2 expression- HML2 per tissue github.R normalizes the raw counts files using TPM. These TPM counts are then used to generate a ggplot boxplot of total HML-2 expression in each tissue
 graph heatmap of individual HML2 proviruses github.R calculates average expression for each HML-2 provirus in TPM and then plots this data using pheatmap 
 
 Separate HML2 expression based on biological sex_ABV8 github uses the metadata from GTEx to separate each donor based on biological sex. Here we used expression of Y chromosome genes in our dataset to double check the provided biological sex. These data were then used to plot HML-2 counts between sexes.
 HML-2 and age and HML-2 and Hardy both classify HML-2 donors in a similar way based on age group or the hardy score(a classification for how they died)
 
 Individual provirus covariate plotting uses Limma-Voom to calculate significant differences in individual provirus expression for each of the three covariates measured. It then plots a ggplot boxplot for each tissue analyzed
 
 TF motif dendro solo takes the TF matrix generated by FIMO on an alignment of HML-2 LTRs and uses that matrix to generate a dendrogram for analysis of this matrix
    
